### PR TITLE
Tests for the v2 sync API

### PR DESCRIPTION
**Pull request by NegativeMjark**
We need to test the following:

* [x] That we can sync a joined room.
 * [x] That we can sync a room joined before syncing. 7aa0e8f
 * [x] That we can sync incremental changes to a room. 5ab8118
 * [x] That when we join a room it is included in the next incremental sync. 93bbbe4

* [x] That presence is included in a sync.
 * [x] That updates to presence are included in an incremental sync. fe78f61
 * [x] That the user is considered to be online in their own sync if they haven't set a set_presence. 97b12eb
 * [x] That the user is considered to be offline in their own sync if they have set a set_presence=offline. 97b12eb

* [x] That the timeline limits are applied correctly when syncing a room
  * [x] That setting the limit to one returns only one event and is limited. cffee9b
  * [x] That setting the limit to a large number on a new room isn't limited. f97be43
  * [x] That setting the limit on a filtered timeline works. ca05366

* [x] That state is included in an initial sync.
 * [x] That the state of the room is included in the initial sync. b6af799
 * [x] That the delta to the state is included in the incremental sync. bdece04

* [x] That messages sent by the user appear in their timeline.
  * [x] That a message sent before an initial sync appears in the timeline. cffee9b
  * [x] That a message sent after an initial sync appears in the timeline of an incremental sync. 5ab8118
  * [x] That a message sent before an initial sync with a transaction id appears in the timeline. 242f0de
  * [x] That the prev_batch token can be used in the v1 messages API to retrieve messages that were cut off by the limit.  b9095be

* [ ] That filters work.
  * [ ] That presence filters work.
  * [x] That room state filters work. b6af799
  * [x] That room timeline filters work. ca05366
  * [ ] That room ephemeral filters work.

* [ ] That read receipts are included in the ephemeral events.
  * [ ] That read receipts are included in the initial sync.
  * [ ] That read receipts are included in the incremental sync.
  * [ ] That read receipts behave correctly when timeline events are filtered.

* [ ] That typing notifications are included in the ephemeral events.
 * [ ] That typing notifications are included in the initial sync.
 * [ ] That typing notifications are included in the incremental sync.

* [x] That rooms that the user is invited to are included in the sync.
 * [x] That invited rooms appear in the initial sync. 1bef55d
 * [x] That new invited rooms appear in an incremental sync. 46e3fc0
 * [x] That when joining a room that the user was invited to the user receives the complete state for the room in the next incremental sync. 15ae515

* [ ] That rooms that the user has left or been banned from appear in the sync.
 * [ ] That rooms that the user has left appear in an initial sync.
 * [ ] That rooms that the user has been banned from appear in the initial sync.
 * [ ] That rooms that the user has newly left appear in the incremental sync.
 * [ ] That rooms that the user has been newly banned from appear in the incremental sync.

* [ ] That the sync API can be polled for updates using the timeout parameter.
 * [x] That the sync API is woken up by new message events. 27b7f9c
 * [ ] That the sync API is woken up by presence events.
 * [ ] That the sync API is woken up by read receipt events.
 * [ ] That the sync API is woken up by typing events.
 * [ ] That the sync API isn't woken up for an event excluded by the filter.

### CODE REVIEW

File: tests/10apidoc/34room-messages.pl [hunk 1]
```diff
@@ -44,10 +44,17 @@ sub matrix_send_room_message
...
 
-   do_request_json_for( $user,
-      method => "POST",
-      uri    => "/api/v1/rooms/$room_id/send/$type",
+   my $method = "POST";
+   my $uri = "/api/v1/rooms/$room_id/send/$type";
+
+   if (defined $opts{txn_id}) {
+      $method = "PUT";
+      $uri = "$uri/${\ $opts{txn_id} }";
```
Comment (leonerd) [inline tests/10apidoc/34room-messages.pl]: This doesn't need the `${\...}` interpolation hack; hash derefs interpolate nicely. It's just method calls that don't.

    $uri = "$uri/$opts{txn_id}";

File: tests/31sync/00user.pl [hunk 2]
```diff
@@ -0,0 +1,10 @@
+push our @EXPORT, qw( matrix_register_sync_user );
+
+my $counter = 0;
+
+sub matrix_register_sync_user {
```
Comment (leonerd) [inline tests/31sync/00user.pl]: Any particular reason for this new function? The regular `matrix_register_user` now generates serial-numbered user IDs if not supplied; so the only thing you're really saving here is the `with_events => 0`.

File: tests/10apidoc/34room-messages.pl (hunk 1 above)
Comment (NegativeMjark) [inline tests/10apidoc/34room-messages.pl]: Done d8a1e75

File: tests/31sync/00user.pl (hunk 2 above)
Comment (NegativeMjark) [inline tests/31sync/00user.pl]: Done  1a49cf5

File: tests/31sync/02sync.pl [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        my ( $body ) = @_;
+        require_json_keys( $body, qw( rooms presence next_batch ) );
+        require_json_keys( $body->{presence}, qw( events ));
+        require_json_keys( my $rooms = $body->{rooms}, qw( joined invited archived ) );
+    });
+}
+
+test "Can sync",
+    requires => [qw( first_api_client can_create_filter )],
+    provides => [qw( can_sync )],
```
Comment (leonerd) [inline tests/31sync/02sync.pl]: blank line above

File: tests/31sync/03joined.pl [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            ( $room_id ) = @_;
+            matrix_sync( $user, filter => $filter_id )
+        })->then( sub {
+            my ( $body ) = @_;
+            my $room = $body->{rooms}{joined}{$room_id};
+            require_json_keys( $room, qw( event_map timeline state ephemeral ));
+            require_json_keys( $room->{timeline}, qw( events limited prev_batch ));
+            require_json_keys( $room->{state}, qw( events ));
+            require_json_keys( $room->{ephemeral}, qw( events ));
+            require_json_keys( $room->{event_map}, @{$room->{timeline}{events}} );
```
Comment (leonerd) [inline tests/31sync/03joined.pl]: Spacing inside `@{ EXPR }`
https://github.com/matrix-org/sytest/blame/develop/STYLE-GUIDE.rst#L130

File: tests/31sync/04timeline.pl [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+            $room->{event_map}{$event_id}{unsigned}{transaction_id} eq "my_transaction_id"
+                or die "Unexpected transaction id";
+            (not $room->{timeline}{limited})
+                or die "Did not expect timeline to be limited";
+            Future->done(1)
+        })
+    };
+
+
+test "That a filtered timeline reaches its limit",
```
Comment (leonerd) [inline tests/31sync/04timeline.pl]: Remove the `That ` part of the message - it'll be printed as

    Testing if: That a filtered timeline...

File: tests/31sync/04timeline.pl [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+            $room->{event_map}{$event_id}{content}{body} eq "My message"
+                or die "Unexpected message body.";
+            (not $room->{timeline}{limited})
+                or die "Did not expect timeline to be limited";
+            Future->done(1)
+        });
+    };
+
+
+test "That syncing a new room with a large timline limit isn't limited",
```
Comment (leonerd) [inline tests/31sync/04timeline.pl]: Typo "timeline"

File: tests/31sync/05presence.pl [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    check => sub {
+        my ( $http ) = @_;
+        my ( $user_a, $user_b, $filter_id_a, $filter_id_b, $next_a );
+        Future->needs_all(
+            matrix_register_user_with_filter( $http, {} ),
+            matrix_register_user_with_filter( $http, {} ),
+        )->then( sub {
+            ( $user_a, $filter_id_a, $user_b, $filter_id_b ) = @_;
+            # We can't use matrix_create_and_join since that polls the event
+            # stream to check that the user has joined.
```
Comment (leonerd) [inline tests/31sync/05presence.pl]: This issue has come up a few times now. I wonder if that should be an option to `matrix_create_and_join_room()` to not do the event stream

File: tests/31sync/06state.pl [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+            room => {
+                timeline => { types => [] },
+                state => { types => ["a.madeup.test.state"] },
+                ephemeral => { types => [] },
+            },
+            presence => {types => [] },
+        };
+
+        matrix_register_user_with_filter( $http, $filter )->then( sub {
+            ( $user,  $filter_id ) = @_;
```
Comment (leonerd) [inline tests/31sync/06state.pl]: Double space

File: tests/31sync/08polling.pl [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+                    matrix_sync( $user,
+                        filter => $filter_id, since => $next, timeout => 10000
+                    ),
+                    delay( 0.1 )->then( sub {
+                        matrix_send_room_text_message(
+                            $user, $room_id, body => "1"
+                        )
+                    })
+                ),
+                delay( 10 )->then_fail("Timeout waiting for sync")
```
Comment (leonerd) [inline tests/31sync/08polling.pl]: This line isn't necessary; the entire test has a 10 second deadline anyway applied by `run-tests.pl` itself

**Comment by leonerd** [comment]
Overall comments: Appears to be going in a good direction. Take a moment to read over the `STYLE-GUIDE.rst` though, as there's various whitespace related mistakes here; mostly small things like insufficient spacing inside circumfix operators.

Don't be afraid of adding extra blank lines inside blocks of code, to separate out the stages. In particular, a blank line after the line that unpacks `@_` into new variables would be good, as it would visually separate it from subsequent declarations of new variables that aren't initially assigned.

File: tests/31sync/03joined.pl (hunk 4 above)
Comment (NegativeMjark) [inline tests/31sync/03joined.pl]: Done df08754

File: tests/31sync/06state.pl (hunk 8 above)
Comment (NegativeMjark) [inline tests/31sync/06state.pl]: Done df08754

File: tests/31sync/02sync.pl (hunk 3 above)
Comment (NegativeMjark) [inline tests/31sync/02sync.pl]: Done c8d24b3

File: tests/31sync/04timeline.pl (hunk 6 above)
Comment (NegativeMjark) [inline tests/31sync/04timeline.pl]: Done c8d24b3

File: tests/31sync/08polling.pl (hunk 9 above)
Comment (NegativeMjark) [inline tests/31sync/08polling.pl]: Done 174be60

File: tests/31sync/04timeline.pl (hunk 5 above)
Comment (NegativeMjark) [inline tests/31sync/04timeline.pl]: Done c8d24b3

**Comment by NegativeMjark** [comment]
I used the following for finding the formatting issues you brought up. You might find it useful if you want to fix the rest of the tests in sytest.

    grep -rnB2 ^test tests | grep '  *}'; grep -rn Then tests; grep -Ern '@{\$|}}' tests;

**Comment by leonerd** [comment]
Having performed some style-related updates, this now LGTM.

Does it have to wait for a similar merge in synapse first?

**Comment by leonerd** [comment]
Oh - it seems not. It runs against develop synapse just fine.

**Pull request merged**